keps/sig-api-machinery/5000-api-linting-crd-schema-tooling/README.md

```diff
@@ -0,0 +1,959 @@
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+This proposal introduces two new sig-api-machinery subprojects to improve the experience of developing Kubernetes APIs (including CRDs).`kube-api-linter` (aka `kal`) is a golangci-lint plugin for evaluating Go types that are used to generate CRDs to ensure they follow the Kubernetes API conventions and best practices. `crdify` is a CLI that compares CRD YAML to identify changes in CRD schemas, ensuring that any changes are compatible.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Despite there existing documentation on [Kubernetes API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md) and [making changes to Kubernetes APIs](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md), there is little to no tooling that enable developers to ensure they are following best practices when developing Kubernetes-native APIs.
+
+This KEP aims to improve the Kubernetes-native API development experience by adding two sig-api-machinery subprojects for:
+
+- Linting Kubernetes APIs written in Go based on the [Kubernetes API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md)
+- Validating changes to CustomResourceDefinition schemas in YAML
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Create tooling to help developers writing Kubernetes APIs in Go to follow best practices and Kubernetes API conventions
+- Create tooling to help developers writing CustomResourceDefinitions avoid making breaking changes
```

Comment:
There has been discussion about potentially an admission plugin version of CRD schema upgrade checks too, is that part of our goals initially?

Reply:
I figured this would be best suited to a KEP that is specifically scoped to that work after the initial tooling is in place and has some time to be hardened a bit.

Reply:
I added https://github.com/kubernetes/enhancements/pull/5102/files#diff-41e4d1a3fd8006a9e40e6415cede33325066400913bdaa49808839037eb2fecaR90 - if that is insufficient I can add this elsewhere and provide more detail